### ISSUE TITLE
Crash on glTextureSubImage2Dwith type == GL_FLOAT

### ISSUE THREAD

**Issue by dannypurcell** [LWJGL/lwjgl3#103]
There seems to be an issue causing a crash whenever a call is made to glTextureSubImage2D with the `type` parameter set to GL_FLOAT.
I've tinkered with this a bit to try an find a combination of parameters that passes without the crash and it seems sending GL_FLOAT for the pixel type is the trigger.

According to the [docs](https://www.opengl.org/sdk/docs/man/html/glTexSubImage2D.xhtml) GL_FLOAT should be an acceptable value for glTextureSubImage2D's pixel type but requesting it seems to blow up every time. 

The following is my first guess at some information that may be helpful in sorting this one out. I will be happy to provide any further system details you like. 

Context:
I hit this issue while trying to replicate [Kai's ray tracer](https://github.com/LWJGL/lwjgl3-wiki/wiki/2.6.1.-Ray-tracing-with-OpenGL-Compute-Shaders-%28Part-I%29) using DSA. Where [his code](http://d2g0ezo1t7nqa0.cloudfront.net/tutorials/141208_glsl_compute/rt_glsl_compute_pt1.zip) calls
```
ByteBuffer black = null;
glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA32F, width, height, 0, GL_RGBA, GL_FLOAT, black);
```
I opted to call
```
ByteBuffer pixels = BufferUtils.createByteBuffer(width * height * Float.BYTES);
glTextureSubImage2D(id, 0, 0, 0, width, height, GL11.GL_RGBA, GL11.GL_FLOAT, pixels);
```
since glTextureSubImage2D does not allow null for the `pixels` parameter.

Process output:
```
Connected to the target VM, address: '127.0.0.1:53766', transport: 'socket'
[LWJGL] Version 3.0.0a | Windows 7 | amd64
[LWJGL] MemoryUtil MemoryAccessor: MemoryAccessorUnsafe
[LWJGL] Failed to locate address for GL function glNamedBufferPageCommitmentARB
[LWJGL] [GL] GL_ARB_sparse_buffer was reported as available but an entry point is missing.
[LWJGL] Failed to locate address for GL function wglCopyImageSubDataNV
[LWJGL] [GL] WGL_NV_copy_image was reported as available but an entry point is missing.
Disconnected from the target VM, address: '127.0.0.1:53766', transport: 'socket'

Process finished with exit code -1073741819 (0xC0000005)
```

Nvidia Driver Version: 347.88

Reproduction Test:
The following junit4 test should only require imports for `org.junit.Test`, `org.lwjgl`, and core java packages. Please let me know if there are any issues running it.

```
public class TextureLoadCrashTest {

    private static final File libraryPath = new File(String.join(File.separator,
                                                                 System.getProperty("java.io.tmpdir"),
                                                                 TextureLoadCrashTest.class.getName().split("\\.")[0]));
    private static final List<File> nativeDependencies = new ArrayList<>();

    static {
        System.setProperty("org.lwjgl.util.Debug", "true");
        nativeDependencies.add(new File(libraryPath, "liblwjgl.dylib"));
        nativeDependencies.add(new File(libraryPath, "liblwjgl.so"));
        nativeDependencies.add(new File(libraryPath, "liblwjgl32.so"));
        nativeDependencies.add(new File(libraryPath, "libopenal.dylib"));
        nativeDependencies.add(new File(libraryPath, "libopenal.so"));
        nativeDependencies.add(new File(libraryPath, "libopenal32.so"));
        nativeDependencies.add(new File(libraryPath, "lwjgl.dll"));
        nativeDependencies.add(new File(libraryPath, "lwjgl32.dll"));
        nativeDependencies.add(new File(libraryPath, "OpenAL.dll"));
        nativeDependencies.add(new File(libraryPath, "OpenAL32.dll"));
        libraryPath.mkdirs();
        System.setProperty("java.library.path", libraryPath.getAbsolutePath());
        try {
            for (File nativeDependency : nativeDependencies) {
                try (ByteArrayOutputStream out = new ByteArrayOutputStream();
                     InputStream inputStream = getResource(nativeDependency.getName()).openStream();
                     FileOutputStream fos = new FileOutputStream(nativeDependency)) {
                    int next;
                    while ((next = inputStream.read()) != -1) {
                        out.write(next);
                    }
                    fos.write(out.toByteArray());
                    fos.flush();
                }
                nativeDependency.deleteOnExit();
            }
            libraryPath.deleteOnExit();
            Field field = ClassLoader.class.getDeclaredField("sys_paths");
            field.setAccessible(true);
            field.set(null, null);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private static URL getResource(String resourceName) {
        ClassLoader loader = GLTest.class.getClassLoader();
        URL url = loader.getResource(resourceName);
        assert url != null : String.format("resource %s not found.", resourceName);
        return url;
    }

    @Test
    public void testCrashOnLoadTexture() {
        int width = 1024;
        int height = 768;
        glfwInit();
        glfwDefaultWindowHints();
        glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
        glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);
        glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 4);
        glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 5);
        glfwWindowHint(GLFW_VISIBLE, GL_FALSE);
        glfwWindowHint(GLFW_RESIZABLE, GL_TRUE);
        glfwWindowHint(GLFW_OPENGL_DEBUG_CONTEXT, GL_TRUE);
        long windowId = glfwCreateWindow(width, height, "TextureLoadCrashTest", NULL, NULL);
        glfwMakeContextCurrent(windowId);
        GLContext.createFromCurrent();
        glEnable(GL_DEBUG_OUTPUT);
        glfwShowWindow(windowId);

        ByteBuffer pixels = BufferUtils.createByteBuffer(width * height * Float.BYTES);
        int id = glCreateTextures(GL11.GL_TEXTURE_2D);
        glTextureStorage2D(id, 1, GL30.GL_RGBA32F, width, height);
        //Switching the pixel type to UNSIGNED_BYTE passes
        glTextureSubImage2D(id, 0, 0, 0, width, height, GL11.GL_RGBA, GL11.GL_FLOAT, pixels);
    }
}
```

**Comment by httpdigest** [comment]
You are feeding glTextureSubImage2D too few data. Please note that RGBA32F requires 128 bytes per texel

**Comment by dannypurcell** [comment]
Thanks for the detailed reply! 
It figures I would have missed something silly like that. I thought I had that angle covered in trying different internal formats but the others probably either made the buffer more than enough or just right.

Since the native side is reading from a pointer location that has too little memory, the crash should be due to an access violation. Unfortunately, the only place this shows up is in the Windows Application logs as a crash report for java with Exception Code: `0xc0000005` which denizens of the internets report can be triggered by access violations. 
From the user side the lack of reporting reduces the debugging options to research, guess, and check. (when the user is a newb like me it seems like the error could be anywhere :sweat_smile:)

What do you think of adding a size check, guarded by `if ( LWJGLUtil.CHECKS )`, to the functions that take ByteBuffers such that hard crashes are preempted by a good exception message?
I can add these checks to my own middle-ware of course but this seems valuable enough to have in the base library and the functions in scope pretty universally have the guarded `GLChecks.ensureBufferObject` already.

Regarding DSA texture initialization:
Thanks again for the info! Your examples, docs, and comments are seriously speeding up the learning process for me.
To be honest, I've been using those functions in pairs as part of a one shot texture creation method in my middle-ware library. I'll probably break them up again if the initial state of the texture space can be used right away.
Also, I found `ByteBuffer black = null;` very intuitive actually. When I got the exception blocking null for `glTextureSubImage2D` it also made sense that zero-ing a sub-section of a texture space would not have been as easy to implement as doing the whole space at once.

**Comment by Spasi** [comment]
> What do you think of adding a size check, guarded by if ( LWJGLUtil.CHECKS ), to the functions that take ByteBuffers such that hard crashes are preempted by a good exception message?
I can add these checks to my own middle-ware of course but this seems valuable enough to have in the base library and the functions in scope pretty universally have the guarded GLChecks.ensureBufferObject already.

The problem is that you cannot cover everything in the API without tons of code and nightmarish maintenance. Just for textures, think about how many formats and types there are, how many more are added with new extensions, how old constraints are being relaxed as new hardware is released. 

Then there's related state like GL_UNPACK_* and GL_PACK_* that further complicate things. You can't have 100% crash-free guarantees without basically writing a small OpenGL driver. I'm afraid this is outside the scope of LWJGL.

**Comment by dustContributor** [comment]
You should use KHR_debug_output and/or ARB_debug_output for these kinds of things. Driver might let you know something. LWJGL provides a handy utility method for doing so:

'''
// So we can set a breakpoint inside the callback.
GL11.glEnable( KHRDebug.GL_DEBUG_OUTPUT_SYNCHRONOUS );
// Now setup and pass the callback to GL.
Closure debugCallback = GLUtil.setupDebugMessageCallback();
'''

Its really really handy when you have some "invalid param" GL error somewhere and you can't figure out what call is causing it. Put a conditional breakpoint inside the callback, wait for the error and bam, actually usable stack trace :smile:

**Comment by dannypurcell** [comment]
I've actually been using that through GL43. I have my initialization code set up to detect if assertions are enabled and configuration all that debugging and checking I could find if they are.
Unfortunately for this case, the OpenGL implementers at NVidia decided to forego the boundary check (likely due to the performance cost). So, all we get is a hard crash.

Great idea about grabbing the stack from the debug callback! 
I have been just sending the messages to SLF4J without including any stack info, I just figured it wouldn't be useful. I just tried adding a contrived RuntimeException to the Logger messages for GL_DEBUG_SEVERITY_MEDIUM and GL_DEBUG_SEVERITY_HIGH and indeed, we get a very nice stack trace to help shed some light on why that message was cast.
Thanks for the post!
